### ISSUE TITLE
2.1.7.0 binding local volume in rw no longer working

### ISSUE THREAD

**Issue by tyx** [docker/for-win#5318]
### Expected behavior

I expect change on my local fs is propagated to my container with a binded volume

### Actual behavior

I have a docker-compose file with : 
```
        volumes:
            # Comment out the next line in production
            - ${PWD-.}/:/srv/app:rw
```

and when my container is running, change in my local volume is not propagated to my container

On 2.1.6.0, it was working.

### Information
docker inspect
``` 
 "Mounts": [
            {
                "Type": "bind",
                "Source": "/host_mnt/c/Users/timothee.barray/www/oxatis-php",
                "Destination": "/srv/app",
                "Mode": "rw,delegated",
                "RW": true,
                "Propagation": "rprivate"
            },
```

I tried consistency, cached and delegated flag, same results.

I use docker-compose and docker through WSL1.

I already tried a `docker volume prune` (and docker system prune).

Anyone can help ?

Thanks

**Comment by tyx** [comment]
I just tried to use docker-compose 1.25.1-rc1 provided by docker-desktop through power shell and same result. So WSL is not involved apparently.

**Comment by tyx** [comment]
And please note that restarting container make changes appear

**Comment by djs55** [comment]
Thanks for your report. I suspect that Docker is not watching for events on the host properly.

As a workaround could you try running a concurrent container with
```
docker run -it -v <path to sources>:/foo alpine sh
```
Hopefully this will make it enable the event watching on `<path to sources>`.

If this doesn't work, could you upload a diagnostics (from the whale menu -> troubleshoot) and quote the diagnostics id here? Thanks!

**Comment by tyx** [comment]
Hello @djs55 ,

Thanks ! This workaround is working !

btw here is the diag id : C7DBF952-51A0-437A-AC7A-AA60C8B24E74/20191213104653

Hope it helps.

**Comment by sameronline** [comment]
I'm having the same issue running WSL + Docker 2.1.7.0 

whenever new changes are done on the host it does not sync or propagate to the container, unless i  restart containers http://bit.ly/38Luqi5

**Comment by sameronline** [comment]
my diag id: `2D7A7E54-861E-4815-9898-BC22C36CB6D0/20191217155525`

**Comment by apm963** [comment]
I am experiencing this issue after upgrading to Edge 2.1.7.0. `4E818E43-38AC-4E5E-AEAC-4C617B571847/20191218153320`.

The workaround above does not work for me. The symptoms are the same although I'm not using WSL (but am using docker-compose). I'm assuming this is related to the FUSE implementation changes. After a shared file is read in the container, changes to that file which are made on the host do not appear in the container.

Within the docker PHP container:

```
# md5sum path/to/file.php
75d5bad5d5efca6958b058f5cef61918  path/to/file.php
```

On my host:

```
$ md5sum path/to/file.php
8317c52a581e3e57a849132154e06637 *path/to/file.php
```

**Comment by tyx** [comment]
To make the workaround working, I need to down my docker-compose stack before to run the simple container with the binding volume. Then I can up my docker-compose stack and see my local changes

**Comment by apm963** [comment]
This is unusual. I am trying again about an hour later and I am able to consistently make file changes on my host and see the file updates are being properly reflected within the containers. One quirk I noticed is it is not instant - sometimes it takes one or two seconds after saving before `md5sum` has an updated value in the container.

**Comment by mikeparker** [comment]
This looks like a real bug, we will fix ASAP. Thanks for the report!

**Comment by sameronline** [comment]
Thanks @mikeparker it looks like it's happening specially with folders not files.

Deleting a folder from host machine does not get it deleted from the container. (no sync)

**Comment by mikeparker** [comment]
Thanks for the details @sameronline i'll attach those details to our internal bug report.

**Comment by djs55** [comment]
Sorry for the delay getting back to you. I've been investigating this but haven't been able to reproduce it on my machine. Could you tell me more about your repro? I have the following questions (sorry it's a long list, but I'm keen to know more about what happened and I can't pin it down at the moment)

In an ideal world, if you could share a full repro with me that would be great. If not:
- @tyx: did it always reproduce?
- @tyx: did it reproduce quickly? (@apm963 reports a delay)
- could you share your `docker-compose.yml` (or something with the same "shape" i.e. with the images and directory names changed consistently?)
- are you starting and stopping containers frequently, or are they long-running?
- are you using `--restart=always` to make containers start on Docker start?
- when a file is out-of-sync in the container, did you notice if it stayed that way even if you edited it again on the host? Or was it only "temporarily" out-of-sync until the next event forced a refresh?
- how are you modifying the shared files? Are you editing the source code on the host and then running them in a container?
- do you ever modify the same file on the host and in the container? This can definitely cause stale cache entries -- we are assuming that source code is edited on the host and only ever read in the container.

So far I've
- found a minor bug when files are bind-mounted rather than directories. I don't think this can cause your symptoms
- tidied up the logging to be clearer about which paths are being watched on the host
- found a small race window when a container is created but the watches are added asynchronously -- I plan to fix this -- but I doubt this causes your symptoms unless you're creating and destroying containers while simultaneously editing files on the host (so not if the containers are long-running)
- checked that it's not an issue to do with case insensitive paths

Thanks for your help so far!

**Comment by tyx** [comment]
Hi @djs55 !

Here is a small reproducer that follow the most my original setup
https://github.com/tyx/docker-win-bug

So I run
- docker up
- I go on localhost:8000 and See Symfony welcome page
- Update the public/index.php with echo "coucou"; exit;
- I can see the message "coucou" on browser
- I remove this echo statement
- I still see it...

And the Docker stuff to build the oxatis/php-fpm-symfony image (private repo)
https://gist.github.com/tyx/d81e022b8c56fa1dedac6725a074e133

> did it always reproduce?

I confirm I always get the issue. And the workaround stills working.

> did it reproduce quickly? (@apm963 reports a delay)

Yes ! 

 > are you starting and stopping containers frequently, or are they long-running?

long-running, I could spent days before stopping them
 
> are you using --restart=always to make containers start on Docker start?

Nop

>  when a file is out-of-sync in the container, did you notice if it stayed that way even if you edited it again on the host? Or was it only "temporarily" out-of-sync until the next event forced a refresh?

Not sure to get what you mean. The file is never forced, no next event
    
> how are you modifying the shared files? Are you editing the source code on the host and then running them in a container?

I edit the source on the host, and thanks to volume I refresh my browser to see my modifcation


>    do you ever modify the same file on the host and in the container? This can definitely cause stale cache entries -- we are assuming that source code is edited on the host and only ever read in the container.

No, I can reproduce the issue with many and many files.

Hope that helps

**Comment by djs55** [comment]
@tyx thanks very much for making a self-contained example and answering my long list of questions, I really appreciate it.

Unfortunately I still can't reproduce it on my machine :( Could you sanity-check what I'm doing just in case? Your instructions are clear, but I thought I'd check anyway.

In a powershell window I have
```
git clone git://github.com/tyx/docker-win-bug
cd docker-win-bug
mkdir public
docker-compose up
```
I wait for the log lines
```
app_1        | [10-Jan-2020 10:21:37] NOTICE: fpm is running, pid 1
app_1        | [10-Jan-2020 10:21:37] NOTICE: ready to handle connections
```
and then open `localhost:8000` in the browser-- I see the welcome page. I use notepad to open the `public/index.php` file and replace all the text with
```
<?php


echo "coucou"; exit;
```
I then hit refresh in the browser and I see "coucou" as expected (but not as I hoped)

Could I ask one more favour? I have an experimental build with several fixes including
- fix a bug where cache invalidation fails because Linux is confused by Windows filename case-sensitivity (the file `filename` is cached but the invalidation is for `FiLeNaMe` which Windows thinks is the same but Linux disagrees)
- fix a race where cache invalidation messages are missed when a container is initially started
- fix a bug where we try to watch a file path for changes: on Windows this is invalid and we must watch the parent directory
- add logs around important events (e.g. "watching path", "unwatching path", "error watching path") and remove confusing log spam

Could you see if this experimental build helps at all? https://download-stage.docker.com/win/edge/41944/Docker%20Desktop%20Installer.exe

If not, could you reproduce the problem (i.e. edit the file, refresh the page) and then upload a fresh set of diagnostics? The diagnostics contain 2 things:
1. `%LOCALAPPDATA%\Docker\log.txt` contains logs like
```
[10:46:14.065][GoBackendProcess  ][Info   ] time="2020-01-10T10:46:14Z" msg="unwatching path C:\\Users\\djs\\bugs\\docker-win-bug"
[10:46:14.065][GoBackendProcess  ][Info   ] time="2020-01-10T10:46:14Z" msg="invalidating all caches"
```
2. a trace of the last 1000 FUSE messages which shows the cache invalidations.

**Comment by tyx** [comment]
@djs55 It stills not working with this 2.1.8.0. 

Here is my diag id : C7DBF952-51A0-437A-AC7A-AA60C8B24E74/20200110133219

So I try to look into the log.txt and I found lines that looks a problem
> [14:30:23.897][GoBackendProcess  ][Info   ] time="2020-01-10T14:30:23+01:00" msg="rejecting access to file using non-canonical name DEBUGCLASSLOADER.PHP (canonical name is DebugClassLoader.php)"

> [14:29:44.495][VpnKitBridge      ][Info   ] time="2020-01-10T14:29:44+01:00" msg="error CloseWrite to: Le canal de communication est sur le point d'être fermé"
[14:29:44.499][VpnKitBridge      ][Info   ] time="2020-01-10T14:29:44+01:00" msg="error copying: Le canal de communication est sur le point d'être fermé."

> [14:22:11.961][GoBackendProcess  ][Error  ] time="2020-01-10T14:22:11+01:00" msg="wsl unexpose error:Post http://unix/forwards/unexpose/port: open \\\\.\\pipe\\dockerWsl2BootstrapExposePorts: Le fichier spécifié est introuvable."

I will try to take more time to help next week.

**Comment by djs55** [comment]
@tyx thanks for the new diagnostics and all your help so far.

Although the logs have been helpful and I've tracked down a couple of bugs, I've not managed to figure out the main problem :( I know it's a big favour to ask but would you be able to demonstrate the problem to me on a screen sharing session over the next few days? Possibly on Monday? We're hoping to release a new stable version really soon and I'd really like to get to the bottom if this if I can. I can set up a meeting link with [Zoom](https://zoom.us/) (or we could use another tech of your choice)

Failing that, perhaps we could switch to real-time chat on the Docker community slack? You can [sign up here](https://dockercommunity.slack.com/join/shared_invite/enQtNzgzMzcwNDMwMzExLTVmNjY1OGY2ODVlZDA2MjhhY2Y4NTg5NDhhZmViNTk1YjMwOWY4ZWQyN2U2Yzg5YzljMjVhNjg5MTdmNjgwOGU) . I'm `djs55` on there too.

I've just created a new diagnostic tool which might be helpful. It shows all the FUSE operations live:
```
docker run -v /run:/run -it djs55/grpcfuse-control:2.2.0 statistics calls --follow
```
When I edit `index.php` and save it I get a stream of messages including
```
...
21:25:36.405732477,45386,invalidation,NOTIFY,Offset=0 Length=0 Users/djs/bugs/docker-win-bug/public/index.php
21:25:36.405828678,45386,invalidation,NOTIFY,OK
```
-- this shows the cache in the VM has been dropped. When I reload the page I see:
```
21:28:11.904786465,45388,remote,GetAttr,Users/djs/bugs/docker-win-bug/public/index.php
21:28:11.905762665,45388,remote,GetAttr,mode 33261 OK
21:28:11.906054765,45389,remote,Open,Users/djs/bugs/docker-win-bug/public/index.php
21:28:11.906732565,45389,remote,Open,2980 OK
21:28:11.907012765,45390,remote,Read,2980
21:28:11.907513364,45390,remote,Read,2980 Offset=0 Length=27 OK
21:28:11.907680964,45391,remote,Release,2980
21:28:11.908174764,45391,remote,Release,OK
```
-- this shows the new file contents being read. This is how it's supposed to look when it's working, I'd be interested to see how it looks for you.

Thanks again for all your help!

**Comment by mikeparker** [comment]
I've reproduced this using 2.1.7.0 and an editor which saves a backup file at the same time as saving the original file (using Notepad++, it doesn't fail with pure Notepad) but the build djs55 linked (2.1.8.0) fixes this and I can edit the file many times and save even with backups and the refresh works fine.

@tyx the first error message (`rejecting access to file using non-canonical name`) is because Linux is case-sensitive and Windows is not, and your code is referring to a file `DEBUGCLASSLOADER.PHP` which doesn't exist, but the filesystem does have a `DebugClassLoader.php`. When switching to running in Linux containers you must be careful with filename case sensitivity.

@tyx can I just triple check you tried the build djs55 linked multiple times? i.e. changing the file multiple times and refreshing after a second or two. We've fixed a major bug in that build and if its still broken it must be something different.

**Comment by tyx** [comment]
Hi guys !

@djs55 
Unfortunately I will quit my job tomorrow and so online meeting will be difficult but I recorded the problem in video https://drive.google.com/file/d/1lK5uOyyp6HyWjF0ywDDTeInAltITeyxq/view?usp=sharing
As you can see, I edit the file to remove my dump, but it stills appear. I have of course the volume mounted ;)

So I ran your new diagnostic tool. And when I edit my files, in fact, there is no new log. I have just log at start time then nothing when I change any files of my mounted volume

@mikeparker I confirm I use the new 2.1.8.0 for 3 days, and I stop the workaround multiple times to see if anything change. But I still get the issue.

**Comment by djs55** [comment]
Thanks for the update. In particular this comment

> So I ran your new diagnostic tool. And when I edit my files, in fact, there is no new log. I have just log at start time then nothing when I change any files of my mounted volume

This means there are no filesystem events, so we must not be watching the right directories.

In your previous comment
> I confirm I always get the issue. And the workaround stills working.

For some reason the workaround (i.e. running `docker run -v` manually before `docker-compose up`) fixes the problem i.e. makes us watch the correct directory.

I think the problem really has to be in the management of the watched paths, which is where I was initially looking. I'll look there again.

(Even though we haven't fixed it yet, this investigation has caused us to find and fix lots of other related problems, so at least that's a positive outcome!)

**Comment by djs55** [comment]
I've got another couple of builds if anyone would like to try them:

- https://download-stage.docker.com/win/stable/42172/Docker%20Desktop%20Installer.exe : this is the latest stable branch build. It's got all the confirmed fixes in it.

For @tyx in particular I've got some experimental builds to test a couple of theories:

- https://download-stage.docker.com/win/stable/42161/Docker%20Desktop%20Installer.exe : this build has a rewrite of the path watching logic (making it less efficient but more reliable)
- https://download-stage.docker.com/win/stable/42174/Docker%20Desktop%20Installer.exe : this build watches paths in a different order to test a theory that we're confusing the underlying filesystem watching library

If anyone has the time to test these builds, let me know how it goes. If it goes wrong, please upload a fresh set of diagnostics for me to analyse :)

Thanks again for all your help so far!

**Comment by patrick-vandy** [comment]
I'm having this same problem on version 2.2.0.0 (42247), and so is the other developer on Windows at my job. Is this supposed to be fixed in 2.2.0.0?

**Comment by patrick-vandy** [comment]
@djs55 The work around of mounting the volume to another container works for me. The odd things is it continued to pick up file changes after I removed the work around container. However, after running a `docker-compose down` and `docker-compose up -d`, then it was back to not picking up the changes.

**Diagnostics ID:** C1A85C4A-0C66-4411-9320-F1E5BAC88B72/20200124195621

[docker-inspect.txt](https://github.com/docker/for-win/files/4110470/docker-inspect.txt)

**Comment by stephen-turner** [comment]
This is now fixed in 2.2.0.3, so closing this ticket.
